src/frontend/src/crypto/pinIdentity.ts

```diff
@@ -0,0 +1,215 @@
+import { SignIdentity } from "@dfinity/agent";
+import { ECDSAKeyIdentity } from "@dfinity/identity";
+import { z } from "zod";
+
+/** This modules defines the crypto operations for the browser storage Identity.
+ *
+ * The identity is an ECDSA keypair. The keypair is meant to be indirectly stored in browser storage (IndexedDB).
+ * Before being stored, the keypair in encrypted twice (symmetrically).
+ *
+ * The first encryption round uses an AES-GCM key, which may be stored (unencrypted _but_ unextractable) in the
+ * browser.
+ * The second encryption round uses an AES-GCM key, derived (PBKDF2) from a user-defined PIN.
+ *
+ * All the data necessary to recreated & retrieve the keys (salt, initialization vectors, iterations) may also
+ * be stored in the browser.
+ *
+ * The following terms are used:
+ *  * construct/reconstruct: how to originally construct and later re-construct the identity keypair
+ *  * generate: originally generate either an AES key or ECDSA keypair
+ *  * derive: derive a key from e.g. the PIN
+ */
+
+export const PinIdentityMaterial = z.object({
+  /* The ECDSA keypair */
+  publicKey: z.instanceof(CryptoKey),
+  encryptedPrivateKey: z.instanceof(ArrayBuffer),
+  /* any of the NIST approved curves
+   * https://developer.mozilla.org/en-US/docs/Web/API/EcKeyGenParams#instance_properties */
+  namedCurve: z.enum(["P-256", "P-384", "P-521"]),
+
+  /* The first round of encryption with browser key */
+  browserIv: z.instanceof(Uint8Array),
+  browserKey: z.instanceof(CryptoKey),
+
+  /* The second round of encryption with PIN key */
+  pinIv: z.instanceof(Uint8Array),
+  pinSalt: z.instanceof(Uint8Array),
+  pinPbkdfIters: z.number(),
+});
+export type NistEc = PinIdentityMaterial["namedCurve"];
+export type PinIdentityMaterial = z.infer<typeof PinIdentityMaterial>;
+
+export const constructPinIdentity = async ({
+  pin,
+}: {
+  pin: string;
+}): Promise<{
+  identity: SignIdentity;
+  pinIdentityMaterial: PinIdentityMaterial;
+}> => {
+  const browserIv = window.crypto.getRandomValues(new Uint8Array(96));
+  const pinIv = window.crypto.getRandomValues(new Uint8Array(96));
+  const pinSalt = window.crypto.getRandomValues(new Uint8Array(96));
+  const pinPbkdfIters: number = 10000;
```

Comment:
Please double check these values

Reply:
bumped to 10000, we'll see what happens when we do some manual testing before releasing